### DISCUSSION TITLE
AttributeError: 'Timestamp' object has no attribute 'items' - when  y_pred = model.predict_one(x)

### DISCUSSION THREAD

**Discussion by XaviMartiCamarasa** [online-ml/river#1084] [Q&A] [answered]
Hi.
I'm just replaying the example on AirlinePassengers for building a simple nowcasting model, but with my data.
I'm getting this error: AttributeError: 'Timestamp' object has no attribute 'items' - when  y_pred = model.predict_one(x)

TE code is below

```
data_orig = pd.read_csv('C:/Users/jrmr/Master/TFM/PEC3/hungary_chickenpox/hungary_chickenpox.csv')
data_orig['Date'] = pd.to_datetime(data_orig['Date'],format="%d/%m/%Y") # convertimos a fecha
data_orig.set_index('Date', inplace=True)

```

```
#Chickepox 
from river import compose
from river import linear_model
from river import preprocessing


def get_ordinal_date(x):
    return {'ordinal_date': x['month'].toordinal()}


model_pox = compose.Pipeline(
   
    ('scale', preprocessing.StandardScaler()),
    ('lin_reg', linear_model.LinearRegression())
)

from river import metrics
from river import utils
import matplotlib.pyplot as plt


def evaluate_model(model): 

    metric = utils.Rolling(metrics.MAE(), 12)

    dates = []
    y_trues = []
    y_preds = []

    for x, y in df_BUDAPEST.T.iteritems():
     

       # print(x)

        # Obtain the prior prediction and update the model in one go
        y_pred = model.predict_one(x)
        model.learn_one(x, y)

        # Update the error metric
        metric.update(y, y_pred)

        # Store the true value and the prediction
[hungary_chickenpox.csv](https://github.com/online-ml/river/files/9968858/hungary_chickenpox.csv)

        dates.append(x['monthevaluate_model(model_pox)
        y_trues.append(y)
        y_preds.append(y_pred)

    # Plot the results
    fig, ax = plt.subplots(figsize=(10, 6))
    ax.grid(alpha=0.75)
    ax.plot(dates, y_trues, lw=3, color='#2ecc71', alpha=0.8, label='Ground truth')
    ax.plot(dates, y_preds, lw=3, color='#e74c3c', alpha=0.8, label='Prediction')
    ax.legend()
    ax.set_title(metric)

evaluate_model(model_pox)

```
[hungary_chickenpox.csv](https://github.com/online-ml/river/files/9968865/hungary_chickenpox.csv)

**Comment by MaxHalford** [comment]
Hey there. The `x` you pass to `learn_one` and `predict_one` is a dictionary of features. It seems to me you're passing a `datetime` object.

I would suggest rerunning the original example yourself, and printing the objects to get an understanding of them.

**Reply by XaviMartiCamarasa** [reply]
Now I changed a little bit the code and still get another error

```

from river import dummy, stats, metrics
from river import linear_model
from datetime import datetime

model_regresion=linear_model.LinearRegression()

for date, value in df_BUDAPEST.T.iteritems():

  dict_date={}
  dict_date["month"]=date.to_pydatetime()

  print (type(dict_date))
  print (dict_date["month"])
  print("value",value)

  model_regresion.learn_one (dict_date,value)
  y_pred = model_regresion.predict_one(dict_date)
  print(y_pred)
```
![image](https://user-images.githubusercontent.com/117338671/200790324-b8fc7ded-5289-4fea-bb55-0878b65c9044.png)


![image](https://user-images.githubusercontent.com/117338671/200789899-a7103418-7dbb-4ff1-83e8-bb9cbebd002a.png)

**Reply by MaxHalford** [reply]
That's normal. The features you pass to a linear regression must be numbers, like most machine learning algorithm. 

It seems to me the issues you're having are more related to your understanding of the models you're using, more so than River itself. I would suggest taking a step back, and trying to understand the model you're using.

**Reply by XaviMartiCamarasa** [reply]
For Static Regressor should be the same? it worked with datetime!

For the issue I had, I changed the datetime for its ordinal number and worked as expected. Many thanks!

**Reply by smastelini** [reply]
Hi @XaviMartiCamarasa, the `StatisticRegressor` is one of the simplest baselines. It works by completely ignoring the supplied `x` and uses a applies a statistic on `y`. That is why it worked in your example. 

Still, `x` is expected to be a dictionary of features for most River models.

**Comment by XaviMartiCamarasa** [comment, marked as answer]
Understood Max

On Wed, Nov 9, 2022 at 10:24 AM Max Halford ***@***.***>
wrote:

> That's normal. The features you pass to a linear regression must be
> numbers, like most machine learning algorithm.
>
> It seems to me the issues you're having are more related to your
> understanding of the models you're using, more so than River itself. I
> would suggest taking a step back, and trying to understand the model you're
> using.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/online-ml/river/discussions/1084#discussioncomment-4095073>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A37HELZO7XFKYBIPJCAXE6DWHNUTZANCNFSM6AAAAAAR3E5DTE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

-- 



INFORMACIÓ SOBRE PROTECCIÓ DE DADES DE LA UNIVERSITAT OBERTA DE 
CATALUNYA (UOC)

Us informem que les vostres dades identificatives i les 
contingudes en els missatges electrònics i fitxers adjunts es poden 
incorporar a les nostres bases de dades amb la finalitat de gestionar les 
relacions i comunicacions vinculades a la UOC, i que es poden conservar 
mentre es mantingui la relació. Si ho voleu, podeu exercir el dret a 
accedir a les vostres dades, rectificar-les i suprimir-les i altres drets 
reconeguts normativament adreçant-vos a l'adreça de correu emissora o a 
***@***.*** ***@***.***>.

Aquest missatge i qualsevol 
fitxer que porti adjunt, si escau, tenen el caràcter de confidencials i 
s'adrecen únicament a la persona o entitat a qui s'han enviat.

Així 
mateix, posem a la vostra disposició un delegat de protecció de dades que 
no només s'encarregarà de supervisar tots els tractaments de dades de la 
nostra entitat, sinó que us podrà atendre per a qualsevol qüestió 
relacionada amb el tractament de dades. La seva adreça de contacte és 
***@***.*** ***@***.***>.
INFORMACIÓN SOBRE PROTECCIÓN DE DATOS DE 
LA UNIVERSITAT OBERTA DE CATALUNYA (UOC)
Os informamos de que vuestros 
datos identificativos y los contenidos en los mensajes electrónicos y 
ficheros adjuntos pueden incorporarse a nuestras bases de datos con el fin 
de gestionar las relaciones y comunicaciones vinculadas a la UOC, y de que 
pueden conservarse mientras se mantenga la relación. Si lo deseáis, podéis 
ejercer el derecho a acceder a vuestros datos, rectificarlos y suprimirlos 
y otros derechos reconocidos normativamente dirigiéndoos a la dirección de 
correo emisora o a ***@***.*** ***@***.***>.
Este mensaje y 
cualquier fichero que lleve adjunto, si procede, tienen el carácter de 
confidenciales y se dirigen únicamente a la persona o entidad a quien se 
han enviado.
Así mismo, ponemos a vuestra disposición a un delegado de 
protección de datos que no solo se encargará de supervisar todos los 
tratamientos de datos de nuestra entidad, sino que podrá atenderos para 
cualquier cuestión relacionada con el tratamiento de datos. Su dirección de 
contacto es ***@***.*** ***@***.***>.


UNIVERSITAT OBERTA DE 
CATALUNYA (UOC) DATA PROTECTION INFORMATION
Your personal data and the data 
contained in your email messages and attached files may be stored in our 
databases for the purpose of maintaining relations and communications 
linked to the UOC, and the data may be stored for as long as these 
relations and communications are maintained. If you so wish, you can 
exercise your rights to access, rectification and erasure of your data, and 
any other legally held rights, by writing to the sender’s email address or 
***@***.*** ***@***.***>.
This message and, where 
applicable, any attachments are confidential and addressed solely to the 
individual or organization they were sent to.
The UOC has a data protection 
officer who not only supervises the data processing carried out at the 
University, but who will also respond to any questions you may have about 
this data processing. You can contact our data protection officer by 
writing to ***@***.*** ***@***.***>.

**Comment by XaviMartiCamarasa** [comment]
Cool, thanks!

On Wed, Nov 9, 2022 at 11:25 AM Saulo Martiello Mastelini <
***@***.***> wrote:

> Hi @XaviMartiCamarasa <https://github.com/XaviMartiCamarasa>, the
> StatisticRegressor is one of the simplest baselines. It works by
> completely ignoring the supplied x and uses a applies a statistic on y.
> That is why it worked in your example.
>
> Still, x is expected to be a dictionary of features for most River models.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/online-ml/river/discussions/1084#discussioncomment-4095554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A37HELYINLBIP2NUKGHMIFDWHN33FANCNFSM6AAAAAAR3E5DTE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 



INFORMACIÓ SOBRE PROTECCIÓ DE DADES DE LA UNIVERSITAT OBERTA DE 
CATALUNYA (UOC)

Us informem que les vostres dades identificatives i les 
contingudes en els missatges electrònics i fitxers adjunts es poden 
incorporar a les nostres bases de dades amb la finalitat de gestionar les 
relacions i comunicacions vinculades a la UOC, i que es poden conservar 
mentre es mantingui la relació. Si ho voleu, podeu exercir el dret a 
accedir a les vostres dades, rectificar-les i suprimir-les i altres drets 
reconeguts normativament adreçant-vos a l'adreça de correu emissora o a 
***@***.*** ***@***.***>.

Aquest missatge i qualsevol 
fitxer que porti adjunt, si escau, tenen el caràcter de confidencials i 
s'adrecen únicament a la persona o entitat a qui s'han enviat.

Així 
mateix, posem a la vostra disposició un delegat de protecció de dades que 
no només s'encarregarà de supervisar tots els tractaments de dades de la 
nostra entitat, sinó que us podrà atendre per a qualsevol qüestió 
relacionada amb el tractament de dades. La seva adreça de contacte és 
***@***.*** ***@***.***>.
INFORMACIÓN SOBRE PROTECCIÓN DE DATOS DE 
LA UNIVERSITAT OBERTA DE CATALUNYA (UOC)
Os informamos de que vuestros 
datos identificativos y los contenidos en los mensajes electrónicos y 
ficheros adjuntos pueden incorporarse a nuestras bases de datos con el fin 
de gestionar las relaciones y comunicaciones vinculadas a la UOC, y de que 
pueden conservarse mientras se mantenga la relación. Si lo deseáis, podéis 
ejercer el derecho a acceder a vuestros datos, rectificarlos y suprimirlos 
y otros derechos reconocidos normativamente dirigiéndoos a la dirección de 
correo emisora o a ***@***.*** ***@***.***>.
Este mensaje y 
cualquier fichero que lleve adjunto, si procede, tienen el carácter de 
confidenciales y se dirigen únicamente a la persona o entidad a quien se 
han enviado.
Así mismo, ponemos a vuestra disposición a un delegado de 
protección de datos que no solo se encargará de supervisar todos los 
tratamientos de datos de nuestra entidad, sino que podrá atenderos para 
cualquier cuestión relacionada con el tratamiento de datos. Su dirección de 
contacto es ***@***.*** ***@***.***>.


UNIVERSITAT OBERTA DE 
CATALUNYA (UOC) DATA PROTECTION INFORMATION
Your personal data and the data 
contained in your email messages and attached files may be stored in our 
databases for the purpose of maintaining relations and communications 
linked to the UOC, and the data may be stored for as long as these 
relations and communications are maintained. If you so wish, you can 
exercise your rights to access, rectification and erasure of your data, and 
any other legally held rights, by writing to the sender’s email address or 
***@***.*** ***@***.***>.
This message and, where 
applicable, any attachments are confidential and addressed solely to the 
individual or organization they were sent to.
The UOC has a data protection 
officer who not only supervises the data processing carried out at the 
University, but who will also respond to any questions you may have about 
this data processing. You can contact our data protection officer by 
writing to ***@***.*** ***@***.***>.
